nifi-nar-bundles/nifi-slack-bundle/nifi-slack-processors/src/test/java/org/apache/nifi/processors/slack/PutSlackTest.java

```diff
@@ -102,6 +106,81 @@ public void testInvalidIconEmoji() {
     }
 
     @Test
+    public void testInvalidDynamicProperties() {
+        testRunner.setProperty(PutSlack.WEBHOOK_URL, server.getUrl());
+        testRunner.setProperty(PutSlack.WEBHOOK_TEXT, WEBHOOK_TEST_TEXT);
+        PropertyDescriptor dynamicProp = new PropertyDescriptor.Builder()
+                .dynamic(true)
+                .name("foo")
+                .build();
+        testRunner.setProperty(dynamicProp, "{\"a\": a}");
+
+        testRunner.enqueue("{}".getBytes());
+        testRunner.run(1);
+        testRunner.assertTransferCount(PutSlack.REL_FAILURE, 1);
+    }
+
+    @Test
+    public void testValidDynamicProperties() {
+        testRunner.setProperty(PutSlack.WEBHOOK_URL, server.getUrl());
+        testRunner.setProperty(PutSlack.WEBHOOK_TEXT, WEBHOOK_TEST_TEXT);
+        PropertyDescriptor dynamicProp = new PropertyDescriptor.Builder()
+                .dynamic(true)
+                .name("foo")
+                .build();
+        testRunner.setProperty(dynamicProp, "{\"a\": \"a\"}");
+
+        testRunner.enqueue("{}".getBytes());
+        testRunner.run(1);
+        testRunner.assertTransferCount(PutSlack.REL_FAILURE, 0);
+    }
+
+    @Test
+    public void testValidDynamicPropertiesWithExpressionLanguage() {
+        ProcessSession session = testRunner.getProcessSessionFactory().createSession();
+        FlowFile ff = session.create();
+        Map<String, String> props = new HashMap<>();
+        props.put("foo", "\"bar\"");
+        props.put("ping", "pong");
+        ff = session.putAllAttributes(ff, props);
+
+        testRunner.setProperty(PutSlack.WEBHOOK_URL, server.getUrl());
+        testRunner.setProperty(PutSlack.WEBHOOK_TEXT, WEBHOOK_TEST_TEXT);
+        PropertyDescriptor dynamicProp = new PropertyDescriptor.Builder()
+                .dynamic(true)
+                .name("foo")
+                .build();
+        testRunner.setProperty(dynamicProp, "{\"foo\": ${foo}, \"ping\":\"${ping}\"}");
```

Comment:
Are the payloads always key/value pairs? If so I'm thinking the dynamic properties could be the key/value pairs, and the processor could create a JSON payload from them?  If the attachment payload is more flexible, then JSON should be good. Any chance of injection attacks or anything?